quint/src/quintVerifier.ts

```diff
@@ -182,7 +184,76 @@ const grpcStubOptions = {
   oneofs: true,
 }
 
-function loadGrpcClient(dist: ApalacheDist): VerifyResult<AsyncCmdExecutor> {
+async function loadProtoDefViaReflection(): Promise<VerifyResult<ProtoPackageDefinition>> {
+  type ServerReflectionResponseSuccess = {
+    file_descriptor_response: {
+      file_descriptor_proto: Buffer[]
+    }
+  }
+  type ServerReflectionResponseFailure = {
+    error_response: {
+      error_code: number
+      error_message: string
+    }
+  }
+  type ServerReflectionResponse = ServerReflectionResponseSuccess | ServerReflectionResponseFailure
+  type ServerReflectionService = {
+    new (url: string, creds: any): ServerReflectionService
+    ServerReflectionInfo: () => any
+  }
+  type ServerReflectionPkg = {
+    grpc: { reflection: { v1alpha: { ServerReflection: ServerReflectionService } } }
+  }
+
+  // Obtain a reflection service client
+  // To load from HTTP:
+  // const protoJson = protobuf.parse(response.data, grpcStubOptions).root.toJSON()
+  // const packageDefinition = proto.fromJSON(protoJson, grpcStubOptions)
+  const protoPath = require.resolve('./reflection.proto');
+  const packageDefinition = proto.loadSync(protoPath, grpcStubOptions);
+  const reflectionProtoDescriptor = grpc.loadPackageDefinition(packageDefinition) as unknown as ServerReflectionPkg
+  const serverReflectionService = reflectionProtoDescriptor.grpc.reflection.v1alpha.ServerReflection
+  const reflectionClient = new serverReflectionService(APALACHE_SERVER_URI, grpc.credentials.createInsecure())
```

Comment:
So with this change, we're going to have to maintain two grpc clients, one of which is just to fetch the file from our own code to generate the second? This seems a bit round-a-bout to me...

Reply:
How is this effort any different from extracting the proto file from the jar file? 😅

The ideal solution is to compile and public Apalache proto files on npm 😄

Reply:
I would not argue against adding Apalache proto file. I think that is a nice solution.

minor correction, my solution doesn't write to disk. it parses the files in memory. nonetheless, your argument stays valid.

Reply:
just wanted to reiterate the differences between different solution to this.

### Extract from Jar

- download and put jar in a location
- pass the location as `APALACHE_DIST`
- create a temporary directory
- extract the proto file by a spawning a separate `jar` process in the temporary directory
- load the client
- delete the temporary directory

Pro: The easiest solution
Con: User has to download and extract the apalache distribution and then pass it as `APALACHE_DIST`

### Extract from gRPC reflection

- include `reflection.proto` - which is a standard file. shouldn't change in near future
- parse it and create gRPC reflection client
- retrieve `cmdExecutor.proto`
- parse and create `CmdExecutor` client

Pro: No need to include apalache source
Con: Dynamic client creation means no statically type checking Apalache gRPC types

### Include `CmdExecutor.proto` in quint

- include `cmdExecutor.proto` in quint
- parse and create `CmdExecutor` client

Pro: Bypass gRPC reflection
Con: Need to sync `CmdExecutor.proto` from apalache source

### Use Apalache gRPC client

- statically compile and publish gRPC client on npm
- directly import the client as dependency

Pro: No gRPC reflection. No parsing. Everything statically compiled.
Con: Need to maintain Apalache gRPC client on npm

Reply:
Thanks for the analysis. 

I think you are missing the context that the current arrangement is only provisional and part of WIP. There has never been any expectation that end users will have to manually manage the apalache distribution. This is only a temporary state of things to unblock work on the integration in the language level while we design out the UX properly.

In practice, in the current state of things, the user still needs the apalache distribution, because they will be running apalache locally. The reflection is only saving the need to set an environment variable which, as I have noted, will go away with our plans for quint to manage the apalache distribution.

Our current plan is that Quint will manage a compatible apalache distribution on its own, that will be separate from any apalache version users may have installed. It will therefore know where the distribution is located, and none of this should be exposed to users. We can still chose whether we want to extract the proto file from the jar, build it in to the dist, use reflection, or generate a grpc client.

Re

> Con: Need to sync CmdExecutor.proto from apalache source

This con isn't relevant given our actual architecture, afaik: I can't think of any change to the apalache proto files that would break the proto interface but not require changes to our typescrips in quint. The coupling is tight enough that we *have* to keep the code in sync for this kind of thing whether or not it goes thru the proto file. On the other hand, the proto file is extremely minimal, and I also can't think of any changes that are planned and it has not been touched in 9 months.

Here's an option you have not enumerated here: we generate grpc client code and include it in our source, without worrying about an NPM dependency.

We have issues discussing an planning out this whole design track, and any input on this would be welcome.

Reply:
I agree with you Shon. I just improved the current provisional arrangement - that is all :raised_hands: I don't see it to be the final solution under the hood. But it already offers (almost) the same UX with when Quint manages a compatible apalache distribution :slightly_smiling_face:

> Here's an option you have not enumerated here: we generate grpc client code and include it in our source, without worrying about an NPM dependency.

And yes, I missed it. But compiling the grpc client will add extra build steps and dependencies for Quint.